src/server/artifact.ts

```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import fs from 'fs';
+import * as util from 'util';
+
+type SaveCallback = (localPath: string, error?: string) => Promise<void>;
+
+export class Artifact {
+  private _localPath: string;
+  private _unaccessibleErrorMessage: string | undefined;
+  private _finishedCallback: () => void;
+  private _finishedPromise: Promise<void>;
+  private _saveCallbacks: SaveCallback[] = [];
+  private _finished: boolean = false;
+  private _deleted = false;
+  private _failureError: string | null = null;
+
+  constructor(localPath: string, unaccessibleErrorMessage?: string) {
+    this._localPath = localPath;
+    this._unaccessibleErrorMessage = unaccessibleErrorMessage;
+    this._finishedCallback = () => {};
+    this._finishedPromise = new Promise(f => this._finishedCallback = f);
+  }
+
+  finishedPromise() {
+    return this._finishedPromise;
+  }
+
+  localPath() {
+    return this._localPath;
+  }
+
+  async localPathAfterFinished(): Promise<string | null> {
+    if (this._unaccessibleErrorMessage)
+      throw new Error(this._unaccessibleErrorMessage);
+    await this._finishedPromise;
+    if (this._failureError)
+      return null;
+    return this._localPath;
+  }
+
+  saveAs(saveCallback: SaveCallback) {
+    if (this._unaccessibleErrorMessage)
+      throw new Error(this._unaccessibleErrorMessage);
+    if (this._deleted)
+      throw new Error(`File already deleted. Save before deleting.`);
+    if (this._failureError)
+      throw new Error(`File not found on disk. Check download.failure() for details.`);
+
+    if (this._finished) {
+      saveCallback(this._localPath).catch(e => {});
+      return;
+    }
+    this._saveCallbacks.push(saveCallback);
+  }
+
+  async failureError(): Promise<string | null> {
+    if (this._unaccessibleErrorMessage)
+      return this._unaccessibleErrorMessage;
+    await this._finishedPromise;
+    return this._failureError;
+  }
+
+  async delete(): Promise<void> {
+    if (this._unaccessibleErrorMessage)
+      return;
+    const fileName = await this.localPathAfterFinished();
+    if (this._deleted)
```

Comment:
move this check up?

Reply:
If we move this up, there could be a race where another delete() happens during `await localPathAfterFinished()`.